File: wrappers/dotnet/indy-sdk-dotnet-test/NonSecretsTests/SearchTest.cs
using Hyperledger.Indy.NonSecretsApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.NonSecretsTests
{
    [TestClass]
    public class SearchTest : NonSecretsIntegrationTestBase
    {
        [TestMethod]
        public async Task TestWalletSearchWorks()
        {
            await NonSecrets.AddRecordAsync(wallet, type, id, value, tags);

            using (var search = await NonSecrets.OpenSearchAsync(wallet, type, queryEmpty, optionsEmpty))
            {
                var searchRecordsJson = await search.NextAsync(wallet, 1);
                var searchRecords = JObject.Parse(searchRecordsJson);

                var records = (JArray)searchRecords["records"];

                Assert.AreEqual(1, records.Count);

                var expected = JObject.FromObject(new
                {
                    id = id,
                    type = (object)null,
                    value = value,
                    tags = (object)null
                });

                Assert.IsTrue(JValue.DeepEquals(expected, records[0]));
            }
        }

        [TestMethod]
        public async Task TestWalletSearchWorksForOptions()
        {
            await NonSecrets.AddRecordAsync(wallet, type, id, value, tags);

            var options = JsonConvert.SerializeObject(
                new
                {
                    retrieveRecords = true,
                    retrieveTotalCount = false,
                    retrieveType = false,
                    retrieveValue = false,
                    retrieveTags = false
                }
            );
            using (var search = await NonSecrets.OpenSearchAsync(wallet, type, queryEmpty, options))
            {
                var searchRecordsJson = await search.NextAsync(wallet, 1);
                var searchRecords = JObject.Parse(searchRecordsJson);

                var records = (JArray)searchRecords["records"];

                Assert.AreEqual(1, records.Count);

                var expected = JObject.FromObject(
                    new
                    {
                        id = id,
                        type = (object)null,
                        value = (object)null,
                        tags = (object)null
                    }
                );

                Assert.IsTrue(JValue.DeepEquals(expected, records[0]));
            }
        }

        [TestMethod]
        public async Task TestWalletSearchWorksForQuery()
        {
            await NonSecrets.AddRecordAsync(wallet, type, id, value, tags);
            await NonSecrets.AddRecordAsync(wallet, type, id2, value2, tags2);
            await NonSecrets.AddRecordAsync(wallet, type, id3, value2, tags3);

            var query = "{\"tagName1\":\"str2\"}";

            using (var search = await NonSecrets.OpenSearchAsync(wallet, type, query, optionsEmpty))
            {
                var searchRecordsJson = await search.NextAsync(wallet, 3);
                var searchRecords = JObject.Parse(searchRecordsJson);

                var records = (JArray)searchRecords["records"];

                Assert.AreEqual(1, records.Count);

                var expected = JObject.FromObject(new
                {
                    id = id2,
                    type = (object)null,
                    value = value2,
                    tags = (object)null
                });

                Assert.IsTrue(JValue.DeepEquals(expected, records[0]));
            }
        }
    }
}

File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/non_secrets/SearchTest.java
package org.hyperledger.indy.sdk.non_secrets;

import org.json.JSONArray;
import org.json.JSONObject;
import org.junit.Test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;

import org.hyperledger.indy.sdk.JsonObjectSimilar;

public class SearchTest extends NonSecretsIntegrationTest {

	@Test
	public void testWalletSearchWorks() throws Exception {
		WalletRecord.add(wallet, type, id, value, tags).get();

		WalletSearch search = WalletSearch.open(wallet, type, queryEmpty, optionsEmpty).get();

		String searchRecordsJson = search.fetchNextRecords(wallet, 1).get();

		JSONObject searchRecords = new JSONObject(searchRecordsJson);

		JSONArray records = searchRecords.getJSONArray("records");

		assertEquals(1, records.length());

		JSONObject expected = new JSONObject()
				.put("id", id)
				.putOpt("type", JSONObject.NULL)
				.put("value", value)
				.put("tags", JSONObject.NULL);

		assertTrue(JsonObjectSimilar.similar(expected, records.get(0)));

		search.close();
	}

	@Test
	public void testWalletSearchWorksForOptions() throws Exception {
		WalletRecord.add(wallet, type, id, value, tags).get();

		String options = "{" +
				"   \"retrieveRecords\": true,\n" +
				"   \"retrieveTotalCount\": false,\n" +
				"   \"retrieveType\": false,\n" +
				"   \"retrieveValue\": false,\n" +
				"   \"retrieveTags\": false" +
				"}";
		WalletSearch search = WalletSearch.open(wallet, type, queryEmpty, options).get();

		String searchRecordsJson = search.fetchNextRecords(wallet, 1).get();

		JSONObject searchRecords = new JSONObject(searchRecordsJson);

		JSONArray records = searchRecords.getJSONArray("records");

		assertEquals(1, records.length());

		JSONObject expected = new JSONObject()
				.put("id", id)
				.putOpt("type", JSONObject.NULL)
				.put("value", JSONObject.NULL)
				.putOpt("tags", JSONObject.NULL);

		assertTrue(JsonObjectSimilar.similar(expected, records.get(0)));

		search.close();
	}

	@Test
	public void testWalletSearchWorksForQuery() throws Exception {
		WalletRecord.add(wallet, type, id, value, tags).get();
		WalletRecord.add(wallet, type, id2, value2, tags2).get();
		WalletRecord.add(wallet, type, id3, value3, tags3).get();

		String query = "{\"tagName1\":\"str2\"}";
		WalletSearch search = WalletSearch.open(wallet, type, query, optionsEmpty).get();

		String searchRecordsJson = search.fetchNextRecords(wallet, 3).get();

		JSONObject searchRecords = new JSONObject(searchRecordsJson);

		JSONArray records = searchRecords.getJSONArray("records");

		assertEquals(1, records.length());

		JSONObject expected = new JSONObject()
				.put("id", id2)
				.putOpt("type", JSONObject.NULL)
				.put("value", value2)
				.put("tags", JSONObject.NULL);

		assertTrue(JsonObjectSimilar.similar(expected, records.get(0)));

		search.close();
	}
}
